File: business-backend/app/schemas/card_key.py
from typing import Optional, List
from datetime import datetime
from pydantic import BaseModel, Field, validator
import random
import string

# 卡密状态枚举
class CardKeyStatus:
    UNUSED = "unused"
    USED = "used"
    INVALID = "invalid"

# 卡密基础模型
class CardKeyBase(BaseModel):
    credits: int = Field(..., description="积分面值")
    batch_no: Optional[str] = Field(None, description="批次号")
    expired_at: Optional[datetime] = Field(None, description="过期时间")
    remark: Optional[str] = Field(None, description="备注")

# 创建卡密请求模型
class CardKeyCreate(CardKeyBase):
    count: int = Field(..., gt=0, le=1000, description="创建数量")
    
    @validator('count')
    def validate_count(cls, v):
        if v <= 0 or v > 1000:
            raise ValueError('创建数量必须在1-1000之间')
        return v

# 卡密模型（数据库返回）
class CardKey(CardKeyBase):
    id: int
    card_key: str
    created_by: Optional[int] = None
    created_at: datetime
    status: str
    used_at: Optional[datetime] = None
    used_by: Optional[int] = None
    used_by_nickname: Optional[str] = None
    
    class Config:
        from_attributes = True

# 卡密列表查询参数
class CardKeyListParams(BaseModel):
    skip: int = 0
    limit: int = 10
    status: Optional[str] = None
    batch_no: Optional[str] = None
    created_start: Optional[datetime] = None
    created_end: Optional[datetime] = None
    order_by: str = "created_at"
    order_desc: bool = True

# 更新卡密状态请求
class CardKeyStatusUpdate(BaseModel):
    status: str = Field(..., description="状态: unused, used, invalid")

# 卡密激活请求
class CardKeyActivate(BaseModel):
    card_key: str = Field(..., min_length=9, max_length=9, description="卡密")
    
    @validator('card_key')
    def validate_card_key(cls, v):
        if not all(c in string.ascii_uppercase + string.digits for c in v):
            raise ValueError('卡密只能包含大写字母和数字')
        return v

# 卡密生成器
def generate_card_key() -> str:
    """生成9位数字和大写字母组成的卡密"""
    chars = string.ascii_uppercase + string.digits
    return ''.join(random.choice(chars) for _ in range(9))

# 卡密批量创建响应
class CardKeyBatchCreateResponse(BaseModel):
    batch_no: str
    count: int
    credits: int
    expired_at: Optional[datetime] = None 

# 定义用于列表返回的包装模型
class CardKeyListResponse(BaseModel):
    items: List[CardKey]
    total: int
    page: int
    per_page: int
    total_pages: int

File: admin-frontend/src/api/card-key.ts
import request from '@/utils/request'
import type { PageParams } from '@/types/common'

// 卡密对象类型定义
export interface CardKey {
  id: number
  card_key: string
  credits: number
  batch_no?: string
  created_by?: number
  created_at: string
  expired_at?: string
  status: 'unused' | 'used' | 'invalid'
  used_at?: string
  used_by?: number
  used_by_nickname?: string
  remark?: string
}

// 创建卡密请求参数
export interface CreateCardKeyParams {
  credits: number
  count: number
  batch_no?: string
  expired_at?: string
  remark?: string
}

// 创建卡密响应
export interface CreateCardKeyResponse {
  batch_no: string
  count: number
  credits: number
  expired_at?: string
}

// 卡密查询参数
export interface CardKeyQueryParams extends PageParams {
  status?: string
  batch_no?: string
  created_start?: string
  created_end?: string
}

// 卡密状态更新请求
export interface CardKeyStatusUpdate {
  status: 'unused' | 'used' | 'invalid'
}

// 卡密列表响应
export interface CardKeyListResponse {
  items: CardKey[]
  total: number
  page: number
  per_page: number
  total_pages: number
}

// 批量创建卡密
export function createCardKeys(data: CreateCardKeyParams) {
  return request({
    url: '/card-keys',
    method: 'post',
    data
  })
}

// 获取卡密列表
export function getCardKeys(params: CardKeyQueryParams) {
  return request({
    url: '/card-keys',
    method: 'get',
    params
  })
}

// 获取单个卡密详情
export function getCardKey(id: number) {
  return request({
    url: `/card-keys/${id}`,
    method: 'get'
  })
}

// 更新卡密状态
export function updateCardKeyStatus(id: number, data: CardKeyStatusUpdate) {
  return request({
    url: `/card-keys/${id}/status`,
    method: 'put',
    data
  })
}

// 删除卡密
export function deleteCardKey(id: number) {
  return request({
    url: `/card-keys/${id}`,
    method: 'delete'
  })
}
